packages/mui-base/src/SnackbarUnstyled/useSnackbar.ts

```diff
@@ -0,0 +1,199 @@
+import * as React from 'react';
+import { unstable_useEventCallback as useEventCallback } from '@mui/utils';
+import { TransitionProps } from 'react-transition-group/Transition';
+import { UseSnackbarParameters, SnackbarCloseReason } from './useSnackbar.types';
+
+/**
+ * The basic building block for creating custom snackbar.
+ *
+ * Demos:
+ *
+ * - [Snackbar](https://mui.com/base/react-snackbar/)
+ */
+export default function useSnackbar(parameters: UseSnackbarParameters) {
+  const {
+    autoHideDuration = null,
+    disableWindowBlurListener = false,
+    onClose,
+    open,
+    ref,
+    resumeHideDuration,
+  } = parameters;
+
+  const timerAutoHide = React.useRef<ReturnType<typeof setTimeout>>();
+  const [exited, setExited] = React.useState(true);
+
+  React.useEffect(() => {
+    if (!open) {
+      return undefined;
+    }
+
+    /**
+     * @param {KeyboardEvent} nativeEvent
+     */
+    function handleKeyDown(nativeEvent: KeyboardEvent) {
+      if (!nativeEvent.defaultPrevented) {
+        // IE11, Edge (prior to using Bink?) use 'Esc'
+        if (nativeEvent.key === 'Escape' || nativeEvent.key === 'Esc') {
+          // not calling `preventDefault` since we don't know if people may ignore this event e.g. a permanently open snackbar
+          if (onClose) {
+            onClose(nativeEvent, 'escapeKeyDown');
+          }
+        }
+      }
+    }
+
+    document.addEventListener('keydown', handleKeyDown);
+
+    return () => {
+      document.removeEventListener('keydown', handleKeyDown);
+    };
+  }, [exited, open, onClose]);
```

Comment:
Hmm, why is `exited` in the dependency array?

Reply:
Looks like it doesn't need to be. Removing it from Material UI Snackbar passes all the test and snackbar is closed on `escape` keypress. I would still keep it for now. Or should I remove from both here and Material UI Snackbar?

Reply:
Could it be related maybe to the transition? Maybe test out more scenarios with transitions. Would be great to understand why it is here.

Reply:
The `exited` state is to determine whether the react transition component is exited and it is set when `onExited` is triggered.

Whenever the snackbar is closed by `escape` key the `onClose` is triggered by us where the `open` prop of the Snackbar is controlled by the user by it being set to false. (For both transition or non-transition snackbars) 
So if escape is pressed, `open` is false and it will returned from this effect early.
```js
if (!open) {
  return undefined;
}
```

So I can safely say that `exited` state is not needed as a dependency. (Transition or non-transition components). So, I am removing it here.

Reply:
Dismiss on escape was added in #29850